hir/src/component/mod.rs

```diff
@@ -0,0 +1,284 @@
+
+/// A component export
+#[derive(Debug)]
+pub struct ComponentExport {
+    /// The module function that is being exported
+    pub function: FunctionIdent,
+    /// The component(lifted) type of the exported function
+    pub function_ty: LiftedFunctionType,
+    /// The method of calling the function
+    pub invoke_method: FunctionInvocationMethod,
+}
+
+/// A [Component] is a collection of [Module]s that are being compiled together as a package and have exports/imports.
+#[derive(Default)]
+pub struct Component {
+    /// This tree stores all of the modules
+    modules: RBTree<ModuleTreeAdapter>,
+
+    /// A list of this component's imports, indexed by function identifier
+    imports: BTreeMap<FunctionIdent, ComponentImport>,
+
+    /// A list of this component's exports, indexed by export name
+    exports: BTreeMap<FunctionExportName, ComponentExport>,
+}
+
+impl Component {
+    /// Create a new, empty [Component].
+    #[inline(always)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Return a reference to the module table for this program
+    pub fn modules(&self) -> &RBTree<ModuleTreeAdapter> {
+        &self.modules
+    }
+
+    /// Return a mutable reference to the module table for this program
+    pub fn modules_mut(&mut self) -> &mut RBTree<ModuleTreeAdapter> {
+        &mut self.modules
+    }
+
+    /// Returns true if `name` is defined in this program.
+    pub fn contains(&self, name: Ident) -> bool {
+        !self.modules.find(&name).is_null()
+    }
+
+    /// Look up the signature of a function in this program by `id`
+    pub fn signature(&self, id: &FunctionIdent) -> Option<&Signature> {
+        let module = self.modules.find(&id.module).get()?;
+        module.function(id.function).map(|f| &f.signature)
+    }
+
+    pub fn imports(&self) -> &BTreeMap<FunctionIdent, ComponentImport> {
+        &self.imports
+    }
+
+    pub fn exports(&self) -> &BTreeMap<FunctionExportName, ComponentExport> {
+        &self.exports
+    }
+}
+
+/// This struct provides an ergonomic way to construct a [Component] in an imperative fashion.
+///
+/// Simply create the builder, add/build one or more modules, then call `link` to obtain a [Component].
+pub struct ComponentBuilder<'a> {
+    modules: BTreeMap<Ident, Box<Module>>,
+    imports: BTreeMap<FunctionIdent, ComponentImport>,
+    exports: BTreeMap<FunctionExportName, ComponentExport>,
+    entry: Option<FunctionIdent>,
+    diagnostics: &'a miden_diagnostics::DiagnosticsHandler,
+}
+impl<'a> ComponentBuilder<'a> {
+    pub fn new(diagnostics: &'a miden_diagnostics::DiagnosticsHandler) -> Self {
+        Self {
+            modules: Default::default(),
+            entry: None,
+            diagnostics,
+            exports: Default::default(),
+            imports: Default::default(),
+        }
+    }
+
+    /// Set the entrypoint for the [Component] being built.
+    #[inline]
+    pub fn with_entrypoint(mut self, id: FunctionIdent) -> Self {
```

Comment:
Does a `Component` ever have a nominal entrypoint? If so, is that information provided out-of-band, or is it something present in the Wasm component module that was translated?

Presumably command-style Wasm components (i.e. with a `main`-like entry) provide that information somewhere, but I can't recall if we were going to be able to emit that style of component from Rust, or if there was some reason why we always had to compile reactors. In the latter case, I suppose we would provide the entrypoint out-of-band somehow, but I'm not clear on how that happens currently.

Reply:
Yes - https://github.com/0xPolygonMiden/compiler/blob/af72115995c5f34bc6b1e7e11b464383d4871273/frontend-wasm/src/component/parser.rs#L382-L385 . A component might have a start section with an entry function. It might be used in command-style components but we probably would not be able to emit them for the reasons I described in https://github.com/0xPolygonMiden/compiler/pull/79 (see last paragraph in the PR description).

Reply:
The `start` function for a component is more like a global constructor for the component, rather than `main`. We _do_ need to support and track those through the IR, because we must ensure those get called before anything else in the program (but after memory is initialized). We'll need a new attribute for them though, I'll make a note to do that.

Anyway, both commands and reactors can have a `start` function, it is expected that it will be called before anything else in the module is called. The convention for commands (at least as far as LLVM-produced Wasm modules go)  is that they export a "normal" `_start` function and just do their global init there, rather than using the Wasm `start` function. I do think compiling the equivalent of a dynamic library (compiled as a reactor) via LLVM makes use of the "real" `start` function to invoke all of the static constructors present in the program (more of a C++ thing, but there are ways to use them from Rust as well).

In any case, I think the long and short of it is that we'll want to track whether a module was compiled as a command or reactor, and can assume those same conventions (e.g. a command always defines a `_start` function which is the entrypoint), and if we observe the presence of `start` functions, we'll want to build up a topologically-ordered array of those functions, to be called prior to `_start` in the final program (regardless of whether the input module was a command or reactor).

Doesn't require any changes here, but just wanted to make sure we have a plan for how we're going to compile those features.

Reply:
"Global constructor" is a great analogy! Thanks for the detailed explanation. I made https://github.com/0xPolygonMiden/compiler/issues/130 to make sure it does not fall through the cracks.